From 1ed01668166a5185119edb5f0cc3edb5b26331a2 Mon Sep 17 00:00:00 2001
From: Alex Deryskyba <alex@codesnake.com>
Date: Sun, 19 Oct 2014 16:20:33 +0300
Subject: [PATCH] Save settings only if they were modified after the last save

This prevents from multiple saving the same settings and helps to resolve
the issue on Amlogic G18REF TV-boxes when setiings may be lost after a poweroff.

On G18REF When you press the red button on the remote the system receives a signal that power
is about to off. XBMC always writes guisettings.xml before exit, and the same settings
may be written several times from different places in code. But the power gets turned off
before the system completes all shutdown procedures. There may be the case that guisettings.xml
is written half-way and couldn't be read upon next boot, so the XBMC creates a new one with
default settings.

With this fix the settings will be written at exit only once, minimizing the risk of being lost.
---
 xbmc/settings/Settings.cpp | 14 ++++++++++++++
 1 file changed, 14 insertions(+)

diff --git a/xbmc/settings/Settings.cpp b/xbmc/settings/Settings.cpp
index b403b63bdf..c5486af7b8 100644
--- a/xbmc/settings/Settings.cpp
+++ b/xbmc/settings/Settings.cpp
@@ -55,12 +55,15 @@
 #include "utils/Variant.h"
 #include "utils/XBMCTinyXML.h"
 #include "utils/log.h"
+#include "utils/Digest.h"
 #include "view/ViewStateSettings.h"
+#include "Util.h"
 
 #define SETTINGS_XML_FOLDER "special://xbmc/system/settings/"
 
 using namespace KODI;
 using namespace XFILE;
+using KODI::UTILITY::CDigest;
 
 //! @todo: remove in c++17
 constexpr const char* CSettings::SETTING_LOOKANDFEEL_SKIN;
@@ -557,6 +560,17 @@ bool CSettings::Save(const std::string &file)
   if (!Save(root))
     return false;
 
+  // Avoid saving if the settings saved earlier are indetical to the current ones
+  if (CFile::Exists(file))
+  {
+    std::string fileMD5 = CUtil::GetFileDigest(file, KODI::UTILITY::CDigest::Type::MD5);
+    TiXmlPrinter xmlPrinter;
+    xmlDoc.Accept(&xmlPrinter);
+    std::string settingsMD5 = CDigest::Calculate(CDigest::Type::MD5,(xmlPrinter.CStr()));
+    if (fileMD5 == settingsMD5)
+      return true;
+  }
+
   return xmlDoc.SaveFile(file);
 }
 
-- 
2.25.1

